packages/lit-ssr/src/lib/render-lit-html.ts

```diff
@@ -0,0 +1,551 @@
+  useCustomElementInstance?: boolean;
+};
+
+/**
+ * Operation to output an attribute with bindings. Includes all bindings for an
+ * attribute, like an attribute template part or AttributeComitter.
+ */
+type AttributePartOp = {
+  type: 'attribute-part';
+  index: number;
+  name: string;
+  strings: Array<string>;
+  tagName: string;
+  useCustomElementInstance?: boolean;
+};
+
+/**
+ * Operator to create a custom element instance.
+ */
+type CustomElementOpenOp = {
+  type: 'custom-element-open';
+  tagName: string;
+  ctor: {new (): HTMLElement};
+  staticAttributes: Map<string, string>;
+};
+
+/**
+ * Operation to render a custom element's attributes. This is separate from
+ * `custom-element-open` because attribute/property parts go in between and need
+ * to run and be set on the instance before we render the element's final
+ * attributes.
+ */
+type CustomElementAttributesOp = {
+  type: 'custom-element-attributes';
+};
+
+/**
+ * Operation to render a custom element's children, usually its shadow root.
+ */
+type CustomElementChildrenOp = {
+  type: 'custom-element-children';
+};
+
+/**
+ * Operation to close a custom element so that its no longer available for
+ * bindings.
+ */
+type CustomElementClosedOp = {
+  type: 'custom-element-close';
+};
+
+type Op =
+  | TextOp
+  | NodePartOp
+  | AttributePartOp
+  | CustomElementOpenOp
+  | CustomElementAttributesOp
+  | CustomElementChildrenOp
+  | CustomElementClosedOp;
+
+const getTemplate = (result: TemplateResult) => {
+  const template = templateCache.get(result.strings);
+  if (template !== undefined) {
+    return template;
+  }
+  const {html, attrNames} = getTemplateHtml(result.strings, result._$litType$);
+  const ast = parseFragment(html, {
+    sourceCodeLocationInfo: true,
+  }) as DefaultTreeDocumentFragment;
+
+  const ops: Array<Op> = [];
+
+  /* The last offset of html written to the stream */
+  let lastOffset: number | undefined = 0;
+
+  /**
+   * Sets `lastOffset` to `offset`, skipping a range of characters. This is
+   * useful for skipping <slot>s and distributed nodes in flattened mode, or
+   * skipping and re-writting lit-html marker nodes.
+   */
+  const skipTo = (offset: number) => {
+    if (lastOffset === undefined) {
+      throw new Error('lastOffset is undefined');
+    }
+    if (offset < lastOffset) {
+      throw new Error(`offset must be greater than lastOffset.
+        offset: ${offset}
+        lastOffset: ${lastOffset}
+      `);
+    }
+    lastOffset = offset;
+  };
+
+  /**
+   * Creates or appends to a text opcode with a substring of the html from the
+   * `lastOffset` flushed to `offset`.
+   */
+  const flushTo = (offset?: number) => {
+    if (lastOffset === undefined) {
+      throw new Error('lastOffset is undefined');
+    }
+    const previousLastOffset = lastOffset;
+    lastOffset = offset;
+    const value = html.substring(previousLastOffset, offset);
+    const op = getLast(ops);
+    if (op !== undefined && op.type === 'text') {
```

Comment:
Add comment about when an op is expected to exist v. need to be created.

Reply:
Done